### PR TITLE
Improve async loading of signup steps

### PR DESCRIPTION
**Pull request by jsnajdr**
I tried to improve the work done in #32282 by introducing `React.lazy` and full dynamic webpack import of the signup steps. The code is less verbose now.

It's strange that the `signup` chunk got much bigger. I expected exactly the opposite 😕

### CODE REVIEW

**Comment by sgomes** [comment]
Great work, @jsnajdr ! The code is indeed a lot less verbose now!

> It's strange that the `signup` chunk got much bigger. I expected exactly the opposite 😕

Interestingly, it didn't get much bigger when compressed, only in its decompressed form (according to the performance advisor bot, in any case). This makes me suspect that there's a lot of easily compressible boilerplate being added at some level. The 1KB extra after compression doesn't seem to me like an issue at all.

File: client/signup/config/step-components.js [hunk 1]
```diff
@@ -1,193 +1,78 @@
...
+	'site-information-phone': loadStep( 'site-information' ),
+	'site-or-domain': loadStep( 'site-or-domain' ),
+	'site-picker': loadStep( 'site-picker' ),
+	'site-style': loadStep( 'site-style' ),
+	'site-title': loadStep( 'site-title' ),
+	'site-topic': loadStep( 'site-topic' ),
+	'site-type': loadStep( 'site-type' ),
+	survey: loadStep( 'survey' ),
+	'survey-user': loadStep( 'survey-user' ),
+	test: loadStep( 'test-step' ),
```
Comment (sgomes) [inline client/signup/config/step-components.js]: This seems to change the existing behaviour. Perhaps we should preserve the condition and instead write:
```
process.env.NODE_ENV === 'development'	? loadStep( 'test-step' ) : undefined,
```
?

File: client/signup/config/step-components.js [hunk 2]
```diff
@@ -1,193 +1,78 @@
...
-	<AsyncLoad require="signup/steps/plans-atomic-store" placeholder={ null } { ...props } />
-);
-const ReaderLandingStepComponent = props => (
-	<AsyncLoad require="signup/steps/reader-landing" placeholder={ null } { ...props } />
-);
-const LaunchSiteComponent = props => (
-	<AsyncLoad require="signup/steps/launch-site" placeholder={ null } { ...props } />
+const loadStep = memoize( stepName =>
+	React.lazy( () =>
+		import( /* webpackChunkName: "async-load-signup-steps-[request]", webpackInclude: /signup\/steps\/[a-z-]+\/index.jsx$/ */ `signup/steps/${ stepName }` )
```
Comment (sgomes) [inline client/signup/config/step-components.js]: This line is interesting. I didn't realise you could generate a chunk for each one of these but keep them all in their own chunks. Great work! 👍

**Comment by sgomes** [comment]
Took me a while to figure out why there are new chunks being created, but this appears to be because they're dead code that isn't being loaded. This PR creates a new chunk for every file that matches the regular expression, regardless of whether it's loaded, hence the new chunks.

**Comment by jsnajdr** [comment]
> Took me a while to figure out why there are new chunks being created, but this appears to be because they're dead code that isn't being loaded.

Yes, I already have a PR to remove one of them: #32385

File: client/signup/config/step-components.js (hunk 1 above)
Comment (jsnajdr) [reply client/signup/config/step-components.js]: My thinking was that the `test-step` is harmless and as it's async-loaded anyway, doesn't add any unneeded code. So I chose the simplest route.

File: client/signup/config/step-components.js (hunk 2 above)
Comment (jsnajdr) [reply client/signup/config/step-components.js]: We already do something similar for Moment.js locales, where each locale file is bundled into its own chunks.

Packaging whole modules is a bit tricky: I need to specify the full path to the index file of each module. Directories are not checked against the `webpackInclude` regexp, they are only traversed and descended into.

File: client/signup/config/step-components.js (hunk 1 above)
Comment (sgomes) [reply client/signup/config/step-components.js]: 👍 If it doesn't affect what end users see, I'm ok with keeping it.

**Comment by sgomes** [comment]
FYI @ramonjd: this PR, if merged, will reduce the verbosity of the async loading I added in #32282, while switching to the native React implementation of async loading (instead of Calypso's custom one).

It likely will not solve the visual flashing issue, but it may make it easier to solve.

**Comment by jsnajdr** [comment]
> Interestingly, it didn't get much bigger when compressed, only in its decompressed form

I just diffed the webpack output and it turns out that the generated code for dynamic imports is rather verbose. Webpack generates 4 alternative versions for each step module:
```js
[
        "./site-picker": [
                "./client/signup/steps/site-picker/index.jsx",
                "async-load-signup-steps-about~async-load-signup-steps-clone-cloning~async-load-signup-steps-clone-cr~21ea39bc",
                "async-load-signup-steps-site-picker"
        ],
        "./site-picker/": [
                "./client/signup/steps/site-picker/index.jsx",
                "async-load-signup-steps-about~async-load-signup-steps-clone-cloning~async-load-signup-steps-clone-cr~21ea39bc",
                "async-load-signup-steps-site-picker"
        ],
        "./site-picker/index": [
                "./client/signup/steps/site-picker/index.jsx",
                "async-load-signup-steps-about~async-load-signup-steps-clone-cloning~async-load-signup-steps-clone-cr~21ea39bc",
                "async-load-signup-steps-site-picker"
        ],
        "./site-picker/index.jsx": [
                "./client/signup/steps/site-picker/index.jsx",
                "async-load-signup-steps-about~async-load-signup-steps-clone-cloning~async-load-signup-steps-clone-cr~21ea39bc",
                "async-load-signup-steps-site-picker"
        ],
]
```
All these alternatives are valid and must be covered.

**Comment by sgomes** [comment]
> All these alternatives are valid and must be covered.

They are, but I'd really like an option that would normalise to one of them and bail out with an error if another one is received 😕

**Comment by ramonjd** [comment]
> if merged, will reduce the verbosity of the async loading I added in #32282, while switching to the native React implementation of async loading (instead of Calypso's custom one)

Thanks @sgomes and @jsnajdr ! The payoff of reduced payload is huge, so if we have to sprinkle a little UX fairy dust to make things neater then so be it. 😄

File: client/signup/config/step-components.js [hunk 3]
```diff
@@ -1,193 +1,78 @@
...
+	'site-or-domain': loadStep( 'site-or-domain' ),
+	'site-picker': loadStep( 'site-picker' ),
+	'site-style': loadStep( 'site-style' ),
+	'site-title': loadStep( 'site-title' ),
+	'site-topic': loadStep( 'site-topic' ),
+	'site-type': loadStep( 'site-type' ),
+	survey: loadStep( 'survey' ),
+	'survey-user': loadStep( 'survey-user' ),
+	test: loadStep( 'test-step' ),
+	themes: 'theme-selection',
```
Comment (jblz) [inline client/signup/config/step-components.js]: Is this intentionally not calling `loadStep( 'theme-selection' )`

Comment (jsnajdr) [reply client/signup/config/step-components.js]: No, that was not intentional. My search&replace regexp failed here for some reason. Thanks for spotting the bug!

**Comment by jsnajdr** [comment]
> Took me a while to figure out why there are new chunks being created, but this appears to be because they're dead code that isn't being loaded.

I landed a few Signup cleanup PRs during this week, and the result is that this PR no longer adds any new step chunks other than the `test` step :tada:

**Comment by sgomes** [comment]
@jsnajdr Still LGTM 👍 

The code looks good, and comparing this against the previous commit I don't see any meaningful difference in bytes over the wire.

**Comment by klimeryk** [comment]
Just noting that this introduced a regression that @delputnam corrected in https://github.com/Automattic/wp-calypso/pull/32664.

**Pull request merged**